src/Microsoft.ApiDesignGuidelines.Analyzers/Core/PassSystemUriObjectsInsteadOfStrings.cs

```diff
@@ -24,15 +29,191 @@ public abstract class PassSystemUriObjectsInsteadOfStringsAnalyzer : DiagnosticA
+                    return false;
+                }
+
+                // now we make sure we actually have overloads that contains uri type parameter
+                if (!CheckOverloads(model, node, cancellationToken))
+                {
+                    // no overload that contains uri as parameter
+                    return false;
+                }
+
+                // now we do more expensive word parsing to see whether this is really right one to report it.
+                return CheckStringParametersSlow(stringParameters, cancellationToken);
+            }
+
+            private bool CheckOverloads(SemanticModel model, SyntaxNode node, CancellationToken cancellationToken)
+            {
+                var uriType = _uri;
+                foreach (var overload in model.GetMemberGroup(node, cancellationToken).OfType<IMethodSymbol>())
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (overload.Parameters.Any(p => p.Type?.Equals(uriType) == true))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool CheckStringParametersQuick(IEnumerable<IParameterSymbol> stringParameters, CancellationToken cancellationToken)
+            {
+                foreach (var parameter in stringParameters)
+                {
+                    if (CheckStringParameter(parameter, cancellationToken))
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private static bool CheckStringParameter(IParameterSymbol parameter, CancellationToken cancellationToken)
+            {
+                foreach (var word in uriWords)
+                {
+                    cancellationToken.ThrowIfCancellationRequested();
+
+                    if (parameter.Name?.IndexOf(word, StringComparison.OrdinalIgnoreCase) >= 0)
+                    {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+
+            private bool CheckStringParametersSlow(IEnumerable<IParameterSymbol> stringParameters, CancellationToken cancellationToken)
+            {
+                foreach (var parameter in stringParameters)
+                {
+                    if (parameter.Name == null ||
+                        !CheckStringParameter(parameter, cancellationToken))
+                    {
+                        // quick check failed
+                        continue;
+                    }
+
+                    string word;
+                    var parser = new WordParser(parameter.Name, WordParserOptions.SplitCompoundWords);
+                    while ((word = parser.NextWord()) != null)
+                    {
+                        foreach (var uriWord in uriWords)
```

Comment:
Minor: Probably can use a `HashSet<string>` with OrdinalIgnoreCase comparer for `uriWords`. You should also rename the field to `s_uriWords`